### ISSUE TITLE
XML entities things

### ISSUE THREAD

**Issue by ph777** [redhat-developer/vscode-xml#262]
Hello,

I'm using todays vsix to have the last changes to DTDs and XML entities. Everything works great, i.e. typing `&ab` I get a list of XML entities starting with `&ab`. For each entity I can see its string value. 

What I miss is:

1. **Hover.** Hovering mouse over an entity does not show its string value. I have to `Go to definition` to see its value.
2. **String value.** The displayed string value can contain another entity, e.g. 

    ``` xml
    <!ENTITY a "a">
    <!ENTITY b "a&b;">
    ```
    In such a case the displayed string value for entity `&b;` is `a&b;`. Which is good, but it would be really helpful, if there could be the completely interpreted value `ab` displayed too.
3. **Autostart intellisense on backspace.** If you write an entity, e.g. `&foo;`, then you place cursor elsewhere, then go to the end of entity and start deleting characters with the backspace key, the intellisense does not autostart. You have to hit Ctrl+P to make it start.
4. **Entity names with underscore and backspace.** If an entity name contains an underscore like `&foo_bar;` and you start deleting characters from the end of the entity name, then intellisense does not start (even when trying to force it with Ctrl+P) until the prefix of entity name is without underscores. Once there is no underscore character, you are offered the entity names if you press Ctrl+P. It would be nice to have the option of choosing between `&foo_bar;` and `&foo_baz;` by deleting the last 2 characters from `&foo_bar;`, i.e. when `&foo_ba` is there.

     There is no problem if you start typing an entity with underscore from the beginning, e.g. if you type letters `&foo_ba` you get the intellisense. The problem is only when the entity name is complete, the intellisense is inactive and you start deleting the characters with backspace. Then it is impossible to start intellisense when there is at least one underscore left. I hope it's clear ;-)

This might be several issues but I think they are all connected, so I write them as a single issue.

None of the issues is a clear error (maybe except number 4). However, solving them would make my life much happier.

Thank you again for you good work!

**Comment by angelozerr** [comment]
Thanks @ph777 for your feedback! 

I read quickly your issues and it seems that it will not very hard to fix them. For the interpreted entity (2)) I don't know what is the rule for the compute. Have you some information about that?

**Comment by ph777** [comment]
Hi, I'm not sure what you mean by "the rule to compute the interpreted
entity".

As far as I know, you get the string value of the entity by recursively
resolving entities in entity's definition until there are no more
unresolved entities in it. Then you have the fully resolved value od the
entity. That is what I think would be great to see in intelisense and on
hover.

P.

Dne st 20. kvě 2020 18:24 uživatel Angelo <notifications@github.com> napsal:

**Comment by angelozerr** [comment]
> Autostart intellisense on backspace. If you write an entity, e.g. &foo;, then you place cursor elsewhere, then go to the end of entity and start deleting characters with the backspace key, the intellisense does not autostart. You have to hit Ctrl+P to make it start.

I think it's not possible to manage this case. I have played with TypeScript language and the behavior is the same. Perhaps I have not understood your issue?

**Comment by angelozerr** [comment]
> As far as I know, you get the string value of the entity by recursively
resolving entities in entity's definition until there are no more
unresolved entities in it. 

Ok thanks for your information, I need to investigate of it's easy to implement.

**Comment by ph777** [comment]
Thank you

Dne po 25. kvě 2020 11:27 uživatel Angelo <notifications@github.com> napsal:

**Comment by angelozerr** [comment]
@ph777 here the current status of your issue:

> Entity names with underscore and backspace.

https://github.com/eclipse/lemminx/pull/718 was merged yesterday so I think you could try it by installing the last vsix.

> Hover.

I'm working on that https://github.com/eclipse/lemminx/pull/726#issuecomment-634602847

**Comment by ph777** [comment]
Thanks, now entity names with underscore and backspace are working.

Somehow, no external entities from DTDs (either declared in DOCTYPE, or imported via %dtd;) are offered by intellisense in this last VSIX. I haven't tried to look for commit when it disappeared. I know they worked before.

**Comment by angelozerr** [comment]
Could you add a simple XML / DTD sample please.

**Comment by angelozerr** [comment]
> I know they worked before.

I'm sorry I don't understand what it is broken? Completion and hyperlink should work for local, external entities. We have tests (your usecase) for that which is working. Could you please give me a sample which doesn't work please. Thanks

**Comment by ph777** [comment]
Good point. Trying that I have found out that is much trickier than I thought.

I have the following DTD:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!ENTITY xxx "XXX">

<!-- Catalog flags {{{ -->
<!ENTITY cf_dropdown           "2"> <!-- jako dropdown -->
<!ENTITY cf_privileges         "4"> <!-- podle pristupovych prav -->
<!ENTITY cf_links              "8"> <!-- vcetne linku -->
<!ENTITY cf_cur_user           "16"> <!-- 0x0010: dovolit filtr podle akt. usera -->
<!ENTITY cf_cur_part           "32"> <!-- 0x0020: dovolit filtr podle akt. partnera -->
<!ENTITY cf_l10n               "64"> <!-- 0x0040: lokalizované stringy -->
<!ENTITY cf_sums               "4096"> <!-- u tabparamu zobrazit sumy / prumery -->
<!ENTITY cf_revision           "8192"> <!-- zpristupnit data revize -->
<!ENTITY cf_flatten_tree       "16384"> <!-- 0x004000 -->
<!ENTITY fdcuf_flatten_tree    "16384"> <!-- 0x004000 -->
<!ENTITY cf_choosable_only     "65536"> <!-- 0x010000 -->
<!ENTITY fdcuf_choosable_only  "65536"> <!-- 0x010000 -->
<!ENTITY cf_cur_role           "131072"> <!-- 0x20000: dovolit filtr podle akt. role -->
<!ENTITY cf_currole_flt        "131072">  <!-- 0x020000 -->
<!ENTITY fdcuf_currole_flt     "131072">  <!-- 0x020000 -->
<!ENTITY cf_download_file      "524288">  <!-- 0x080000: CTLG: Add download file button to edit dialog -->
<!ENTITY cf_edit               "4194304"> <!-- 0x400000: CTLG: Add edit button to edit dialog -->
<!ENTITY cf_preview            "8388608"> <!-- 0x800000: CTLG: Add preview button to edit dialog -->
<!ENTITY cf_workflow          "16777216"> <!-- 0x1000000: CTLG: Add workflow data -->
<!ENTITY cf_edit_online       "33554432"> <!-- 0x2000000: CTLG: Add edit_online button to edit dialog -->
<!-- }}} -->
<!-- Tabparam flags {{{ -->
<!ENTITY fdcuf_noedit           "128">  <!-- 0x000080 -->
<!ENTITY fdcuf_nodelete         "256">  <!-- 0x000100 -->
<!ENTITY fdcuf_nonew            "512">  <!-- 0x000200 --> 
<!ENTITY fdcuf_noallatonce      "1024"> <!-- 0x000400 -->
<!ENTITY fdcuf_tab_nodata       "2048"> <!-- 0x000800 -->
<!ENTITY fdcuf_tab_sumrow       "4096"> <!-- 0x001000 -->
<!ENTITY fdcuf_always_inplace   "1048576"> <!-- 0x100000 -->
<!ENTITY fdcuf_hide_actions_column "2097152"> <!-- 0x200000 -->
<!-- }}} -->
```

and xml document

```xml
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE alex-update-sequence SYSTEM "./my.dtd" [<!-- {{{ -->
]>
<!-- }}} -->
<root>
&x|
</root>
```

Only standard XML entities are offered when typing `&x`. It seems like DTD is not parsed.

The funny fact: Try to delete almost any row (even a row with comments, e.g. the first comment, the line with cf_dropdown, the line with cf_privileges, the line with "Tabaparams" comment, the line with fdcuf_noedit, ...) from the DTD, and the entities are offered. I have to say I'm puzzled.

**Comment by angelozerr** [comment]
@ph777 I don't understand your problem, I tried it and I have the proper completion, hyperlink:

![image](https://user-images.githubusercontent.com/1932211/83119345-e3475a00-a0cf-11ea-92b2-40f987cdd3c8.png)

Could you please explain more your problem. Thanks!

**Comment by angelozerr** [comment]
Please note that if your DTD have a syntax error, the entities reference are not retrieved. Perhaps it's your problem?

**Comment by ph777** [comment]
Thanks you that you check it in your environment. But using the exact same data as pasted above, I get

![Screenshot_20200528_111454](https://user-images.githubusercontent.com/20420700/83122967-af226800-a0d4-11ea-81a7-5e024a03ebab.png)

Could this be a difference in the environment? 

LF line endings in XML and DTD.

VSCode
Version: 1.45.1
Commit: 5763d909d5f12fe19f215cbfdd29a91c0fa9208a
Date: 2020-05-14T08:27:22.494Z
Electron: 7.2.4
Chrome: 78.0.3904.130
Node.js: 12.8.1
V8: 7.8.279.23-electron.0
OS: Linux x64 4.4.180-102-default

Java
java version "1.8.0_191"
Java(TM) SE Runtime Environment (build 1.8.0_191-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.191-b12, mixed mode)

**Comment by ph777** [comment]
And I'm using the latest VSIX from https://download.jboss.org/jbosstools/vscode-xml/staging/?C=M;O=D

vscode-xml-0.12.0-4297.vsix

**Comment by angelozerr** [comment]
Could you please create a zip which contains your DTD, XML file and vscode settings please.

**Comment by fbricon** [comment]
@angelozerr I'm seeing the same issue. entities are proposed only once I add a or remove an element/entity/comment from the dtd. This is uncanny

**Comment by angelozerr** [comment]
I cannot reproduce it. @fbricon  If I understand correctly if you open vscode and open completion it doesn't work. If you do some changed in the DTD and reopen completion it works. Is that?

**Comment by fbricon** [comment]
yep
![May-28-2020 12-09-01](https://user-images.githubusercontent.com/148698/83128746-214a7b00-a0dc-11ea-80bb-0da66f4f9068.gif)

There's nothing suspicious in the logs. Nothing

**Comment by fbricon** [comment]
@angelozerr this is where it fails:
<img width="1438" alt="Screen Shot 2020-05-28 at 12 42 16 PM" src="https://user-images.githubusercontent.com/148698/83131822-c36c6200-a0e0-11ea-90e2-df1c031cb5c2.png">

Exceptions are swallowed instead of being logged

**Comment by fbricon** [comment]
When printing the ArrayIndexOutOfBoundsException stacktrace:

```
java.lang.ArrayIndexOutOfBoundsException: Index -9 out of bounds for length 2048
	at org.eclipse.lemminx.extensions.dtd.contentmodel.CMDTDDocument$ScannedDTDEntityDecl.getEntityNameStartColumnNumber(CMDTDDocument.java:165)
	at org.eclipse.lemminx.extensions.dtd.contentmodel.CMDTDDocument$ScannedDTDEntityDecl.createNameParameter(CMDTDDocument.java:134)
	at org.eclipse.lemminx.extensions.dtd.contentmodel.CMDTDDocument$ScannedDTDEntityDecl.<init>(CMDTDDocument.java:108)
	at org.eclipse.lemminx.extensions.dtd.contentmodel.CMDTDDocument.internalEntityDecl(CMDTDDocument.java:276)
	at org.apache.xerces.impl.XMLDTDScannerImpl.scanEntityDecl(Unknown Source)
	at org.apache.xerces.impl.XMLDTDScannerImpl.scanDecls(Unknown Source)
	at org.apache.xerces.impl.XMLDTDScannerImpl.scanDTDExternalSubset(Unknown Source)
	at org.apache.xerces.impl.dtd.XMLDTDLoader.loadGrammar(Unknown Source)
	at org.eclipse.lemminx.extensions.dtd.contentmodel.CMDTDDocument.loadGrammar(CMDTDDocument.java:335)
	at org.eclipse.lemminx.extensions.dtd.contentmodel.CMDTDContentModelProvider.createCMDocument(CMDTDContentModelProvider.java:69)
	at org.eclipse.lemminx.extensions.contentmodel.model.ContentModelManager.findCMDocument(ContentModelManager.java:219)
	at org.eclipse.lemminx.extensions.contentmodel.model.ContentModelManager.findCMDocument(ContentModelManager.java:124)
	at org.eclipse.lemminx.extensions.contentmodel.model.ContentModelManager.findCMDocument(ContentModelManager.java:94)
	at org.eclipse.lemminx.extensions.contentmodel.model.ContentModelManager.findCMDocument(ContentModelManager.java:90)
	at org.eclipse.lemminx.extensions.contentmodel.participants.ContentModelCompletionParticipant.onTagOpen(ContentModelCompletionParticipant.java:71)
	at org.eclipse.lemminx.services.XMLCompletions.collectOpenTagSuggestions(XMLCompletions.java:608)
	at org.eclipse.lemminx.services.XMLCompletions.collectInsideContent(XMLCompletions.java:733)
	at org.eclipse.lemminx.services.XMLCompletions.doComplete(XMLCompletions.java:224)
	at org.eclipse.lemminx.services.XMLLanguageService.doComplete(XMLLanguageService.java:139)
```

**Comment by fbricon** [comment]
More debugging info:
<img width="1081" alt="Screen Shot 2020-05-28 at 1 00 46 PM" src="https://user-images.githubusercontent.com/148698/83133534-4ee6f280-a0e3-11ea-9c5d-247a64a6c096.png">

**Comment by angelozerr** [comment]
Thanks so much @fbricon for your help. I can reproduce it by setting my file to LF (instead of CRLF). I'm fixing that.

**Comment by angelozerr** [comment]
@ph777 please retry to install vscode-xml with last version of vsix which should:

 * fix your bug.
 * provide hover for entities.

Any feedback are welcome!

**Comment by ph777** [comment]
Thank you so much! This is working great!

Any plans on resolving the real value of an entity? We'd really appreciate
it.

<!ENTITY a "A">
<!ENTITY b "&a;B">

Having the above definition the value of "&b;" should be "AB". Now it says
"&a;B" instead. Could this be fixed too?

Thank you for your amazing work once again!

Pavel


út 2. 6. 2020 v 15:13 odesílatel Angelo <notifications@github.com> napsal:

**Comment by angelozerr** [comment]
@ph777 you can install now the release 0.12.0. 

I close this issue since main features has been implemented. Please create an issue per requirement. Thanks.
